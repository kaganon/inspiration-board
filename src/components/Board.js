import React, { Component } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';

import './Board.css';
import Card from './Card';
import NewCardForm from './NewCardForm';
import CARD_DATA from '../data/card-data.json';

class Board extends Component {

  constructor(props) {
    super(props);

    this.state = {
      cards: [],
    };
  }

  componentDidMount() {

    const url = `${this.props.url}/boards/katrina/cards`;

    axios.get(url)
    .then((response) => {
      this.setState({
        cards: response.data,
      });
    })
    .catch((error) => {
      const errors = error.response.data.errors.text;
      this.setState({
        error: errors,
      });
    });
  };


  deleteCard = (id) => {

    const url = `${this.props.url}/cards/${id}`;

    axios.delete(url)
    .then(() => {
      const updatedCardsList = this.state.cards;
      updatedCardsList.forEach((card, i) => {
        if (id === card.card.id) {
          updatedCardsList.splice(i, 1);
        };
      });
      this.setState({
        cards: updatedCardsList
      });
    })
    .catch((error) => {
      const errors = error.response.data.errors.text;
      this.setState({
        error: errors,
      });
    });
  };


  addCard = (cardData) => {

    const url = `${this.props.url}/boards/katrina/cards`;

    axios.post(url, cardData)
    .then((response) => {
      const updatedCardsList = [...this.state.cards, response.data]
      this.setState({
        cards: updatedCardsList,
      });
    })
    .catch((error) => {
      const errors = error.response.data.errors.text;
      this.setState({
        error: errors,
      });
    });
  };


  render() {
    const cardList = this.state.cards.map((card, i) => {
      const cardData = card.card;

      return (
        <Card
          key={i}
          {...cardData}
          onDeleteClickCallback={this.deleteCard}
          />
        )
    });


    const errorMessages = this.state.error ? this.state.error.map((message, i) => {
      return (<h3><li key={i}>{message}</li></h3>) }) : '';


    return (
      <section>
        <header className="validation-errors-display">
          <ul className="validation-errors-display__list">
            {errorMessages}
          </ul>
        </header>
        <div className="board">
          {cardList}
        </div>
        <NewCardForm addCardCallback={this.addCard} />
      </section>
    )
  }

}


Board.propTypes = {
  cards: PropTypes.array,
};

export default Board;
