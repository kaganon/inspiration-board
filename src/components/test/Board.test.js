import React from 'react';
import Board from '../Board';
import { mount, shallow } from 'enzyme';

const CARDS = [
  {
    text: 'cookie cookie nom nom',
    emoji: 'cookie'
  },
  {
    text: 'donut talk to me i am hungry',
    emoji: 'doughnut'
  }

];

describe('Board', () => {
  test('that it matches an existing snapshot', () => {

    // Arrange
    const cards = CARDS;
    const wrapper = shallow( <Board cards={cards} />);

    // Assert
    expect(wrapper).toMatchSnapshot();
  });

  test('renders without crashing even with no cards', () => {

    // Arrange
    const cards = [];
    const wrapper = shallow( <Board cards={cards} />);

    // Assert
    expect(wrapper).toMatchSnapshot();
  });

  test('successfully deep mounts', () => {

    // Arrange
    const wrapper = mount(<Board cards={CARDS} />);

    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
  });

});
