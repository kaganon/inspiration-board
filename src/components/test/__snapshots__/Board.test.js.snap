// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board renders without crashing even with no cards 1`] = `
<section>
  <header
    className="validation-errors-display"
  >
    <ul
      className="validation-errors-display__list"
    />
  </header>
  <div
    className="board"
  />
  <NewCardForm
    addCardCallback={[Function]}
  />
</section>
`;

exports[`Board successfully deep mounts 1`] = `
<Board
  cards={
    Array [
      Object {
        "emoji": "cookie",
        "text": "cookie cookie nom nom",
      },
      Object {
        "emoji": "doughnut",
        "text": "donut talk to me i am hungry",
      },
    ]
  }
>
  <section>
    <header
      className="validation-errors-display"
    >
      <ul
        className="validation-errors-display__list"
      />
    </header>
    <div
      className="board"
    />
    <NewCardForm
      addCardCallback={[Function]}
    >
      <div
        className="new-card-form"
      >
        <h3
          className="new-card-form__header"
        >
          Give me some inspo!
        </h3>
        <div>
          <form
            className="new-card-form__form"
            onSubmit={[Function]}
          >
            <div>
              <label
                className="new-card-form__form-label"
                htmlFor="text"
              >
                Text
              </label>
            </div>
            <textarea
              className="new-card-form__form-textarea"
              name="text"
              onChange={[Function]}
              value=""
            />
            <label
              htmlFor="emoji-select"
            >
              Emoji
            </label>
            <select
              className="new-card-form__form-select"
              id="emoji-select"
              name="emoji"
              onChange={[Function]}
            >
              <option
                key="0"
              />
              <option
                key="1"
              >
                😍
              </option>
              <option
                key="2"
              >
                🍺
              </option>
              <option
                key="3"
              >
                👏
              </option>
              <option
                key="4"
              >
                💖
              </option>
              <option
                key="5"
              >
                😻
              </option>
              <option
                key="6"
              >
                🐶
              </option>
              <option
                key="7"
              >
                💩
              </option>
              <option
                key="8"
              >
                🍪
              </option>
              <option
                key="9"
              >
                🍩
              </option>
            </select>
            <input
              className="new-card-form__form-button"
              name="submit"
              type="submit"
              value="Add Card"
            />
          </form>
        </div>
      </div>
    </NewCardForm>
  </section>
</Board>
`;

exports[`Board that it matches an existing snapshot 1`] = `
<section>
  <header
    className="validation-errors-display"
  >
    <ul
      className="validation-errors-display__list"
    />
  </header>
  <div
    className="board"
  />
  <NewCardForm
    addCardCallback={[Function]}
  />
</section>
`;
